source/grid/grid_tools.cc

```diff
@@ -6647,6 +6648,136 @@ namespace GridTools
+      for (unsigned int v = 0; v < n_vertices; ++v)
+        if (ls_values[mask[v]] < iso_level)
+          configuration |= (1 << v);
+
+      // cell is not cut (nothing to do)
+      if (cut_line_table[configuration] == 0)
+        return;
+
+      // helper function to determine where an edge (between index i and j) is
+      // cut - see also: http://paulbourke.net/geometry/polygonise/
+      const auto interpolate = [&](const unsigned int i, const unsigned int j) {
+        if (std::abs(iso_level - ls_values[mask[i]]) < tolerance)
+          return points[mask[i]];
+        if (std::abs(iso_level - ls_values[mask[j]]) < tolerance)
+          return points[mask[j]];
+        if (std::abs(ls_values[mask[i]] - ls_values[mask[j]]) < tolerance)
+          return points[mask[i]];
+
+        const double mu = (iso_level - ls_values[mask[i]]) /
+                          (ls_values[mask[j]] - ls_values[mask[i]]);
+
+        return Point<dim>(points[mask[i]] +
+                          mu * (points[mask[j]] - points[mask[i]]));
+      };
+
+      // determine the position where edges are cut (if they are cut)
+      std::array<Point<dim>, n_lines> vertex_list_all;
+      for (unsigned int l = 0; l < n_lines; ++l)
+        if (cut_line_table[configuration] & (1 << l))
+          vertex_list_all[l] =
+            interpolate(line_to_vertex_table[l][0], line_to_vertex_table[l][1]);
+
+      // merge duplicate vertices if possible
+      unsigned int                      local_vertex_count = 0;
+      std::array<Point<dim>, n_lines>   vertex_list_reduced;
+      std::array<unsigned int, n_lines> local_remap;
+      std::fill(local_remap.begin(), local_remap.end(), X);
+      for (int i = 0; new_line_table[configuration][i] != X; ++i)
+        if (local_remap[new_line_table[configuration][i]] == X)
+          {
+            vertex_list_reduced[local_vertex_count] =
+              vertex_list_all[new_line_table[configuration][i]];
+            local_remap[new_line_table[configuration][i]] = local_vertex_count;
+            local_vertex_count++;
+          }
+
+      // write back vertices
+      const unsigned int n_vertices_old = vertices.size();
+      for (unsigned int i = 0; i < local_vertex_count; ++i)
+        vertices.push_back(vertex_list_reduced[i]);
+
+      // write back cells
+      if (write_back_cell_data && dim > 1)
+        {
+          for (unsigned int i = 0; new_line_table[configuration][i] != X;
+               i += n_sub_vertices)
+            {
+              cells.resize(cells.size() + 1);
+              cells.back().vertices.resize(n_sub_vertices);
+
+              for (unsigned int v = 0; v < n_sub_vertices; ++v)
+                cells.back().vertices[v] =
+                  local_remap[new_line_table[configuration][i + v]] +
+                  n_vertices_old;
+            }
+        }
+    }
+
+
+
+    template <int dim>
+    void
+    merge_duplicate_points(std::vector<Point<dim>> &vertices)
```

Comment:
I had to define this Function prior to the `process` function so I also moved the `internal::process_sub_cell()` function.

Reply:
> Exactly. Renumber the vertices of the cells. You already have this info, don't you? You could use the same function; if it gets an empty vector of cells, it does nothing.

I tried to include this, see https://github.com/dealii/dealii/pull/14323/commits/587ccb50d3553c9a4ea4f3d724b89c500a146b11. I am not sure about the performance of accessing the `Point<dim>` key of the map for every cell-vertex-point. If I remap the cell vertices, as soon as `create_triangulation` is called, an assert is thrown

```cpp
--------------------------------------------------------
An error occurred in line <11715> of file </home/magdalena/TUM/dealii_work/source/grid/tria.cc> in function
    void dealii::Triangulation<<anonymous>, <anonymous> >::create_triangulation(const std::vector<dealii::Point<spacedim> >&, const std::vector<dealii::CellData<dim> >&, const dealii::SubCellData&) [with int dim = 1; int spacedim = 2]
The violated condition was: 
    !(correct(i, j) ^ (neighbor->direction_flag() == (*cell)->direction_flag()))
Additional information: 
    (none)
--------------------------------------------------------
```
Do you know by chance what we need to consider here regarding the vertex numbering?

Reply:
> [with int dim = 1; int spacedim = 2]

It is a 1D mesh in 2D space, i.e., we have simple lines here and the resulting mesh is not orientatble. This is odd. I don't know when this is the case. Does anyone know and/or have a solution?

For the time being, let's not merge the vertices in the `CellData` case.

Reply:
done!